File: AIOUSB/samples/USB-DA12-8A/Sample.java
/*
 * $RCSfile: Sample.java,v $
 * $Date: 2010/01/29 23:00:04 $
 * $Revision: 1.1 $
 * jEdit:tabSize=4:indentSize=4:collapseFolds=1:
 */

// {{{ build instructions
/*
 * to compile:
 *   javac -cp ../../java/aiousb.jar Sample.java
 * to run:
 *   java -cp ../../java/aiousb.jar:. Sample
 */
// }}}

// {{{ imports
import com.acces.aiousb.*;
import java.io.*;
import java.util.*;
// }}}

class Sample {
	public static final String VERSION = "1.1, 29 January 2010";

	public static void main( String args[] ) {
		USBDeviceManager deviceManager = null;
		try {
/*API*/		deviceManager = new USBDeviceManager();
/*API*/		deviceManager.open();
			System.out.println(
				  "USB-DA12-8A sample program version: " + VERSION + "\n"
/*API*/			+ "  AIOUSB Java library version: " + deviceManager.VERSION_NUMBER + ", " + deviceManager.VERSION_DATE + "\n"
/*API*/			+ "  AIOUSB library version: " + deviceManager.getAIOUSBVersion() + ", " + deviceManager.getAIOUSBVersionDate() + "\n"
				+ "  JRE version: " + System.getProperty( "java.version" ) + "\n"
				+ "  OS version: " + System.getProperty( "os.name" )
					+ " " + System.getProperty( "os.arch" )
					+ " " + System.getProperty( "os.version" ) + "\n"
				+ "  This program demonstrates controlling a USB-DA12-8A device on\n"
				+ "  the USB bus. For simplicity, it uses the first such device found\n"
				+ "  on the bus."
			);
/*API*/		deviceManager.printDevices();
/*API*/		USBDevice[] devices = deviceManager.getDeviceByProductID( USBDeviceManager.USB_DA12_8A_REV_A, USBDeviceManager.USB_DA12_8A );
			if( devices.length > 0 ) {
				USB_DA12_8A_Family device = ( USB_DA12_8A_Family ) devices[ 0 ];	// get first device found
/*API*/			device.setCommTimeout( 1000 );

				/*
				 * demonstrate accessing device properties
				 */
/*API*/			System.out.println( "Found device \'" + device.getName() + "\' with serial number " + Long.toHexString( device.getSerialNumber() ) );

				/*
				 * demonstrate writing D/A counts to one D/A channel
				 */
/*API*/			final int numChannels = device.dac().getNumChannels();
				final int TEST_CHANNEL = 0;		// channels are numbered 0 to numChannels - 1
				final char counts = DA12_AnalogOutputSubsystem.MAX_COUNTS / 2;	// half of full scale
/*API*/			device.dac().writeCounts( TEST_CHANNEL, counts );
				System.out.println( ( int ) counts + " D/A counts successfully output to channel " + TEST_CHANNEL );

				/*
				 * demonstrate writing D/A counts to multiple D/A channels
				 */
				char[] countPoints = new char[ numChannels * 2 ];	// channel/count pairs
				for( int channel = 0; channel < numChannels; channel++ ) {
					countPoints[ channel * 2 ] = ( char ) channel;
					countPoints[ channel * 2 + 1 ] = ( char ) ( ( channel + 1 ) * DA12_AnalogOutputSubsystem.MAX_COUNTS / numChannels );
				}	// for( int channel ...
/*API*/			device.dac().writeCounts( countPoints );
				System.out.println( "Multiple D/A counts successfully output to " + numChannels + " channels" );

				/*
				 * demonstrate writing volts to one D/A channel; note that the hardware range is selected
				 * by means of jumpers; this software setting is merely so that the correct D/A count values
				 * are calculated for the current hardware range setting; this demonstration assumes the
				 * hardware jumpers have been set to the +/-10V range
				 */
				final int HARDWARE_RANGE = DA12_AnalogOutputSubsystem.RANGE_10V;
				final double volts = 5;
/*API*/			device.dac().setRange( HARDWARE_RANGE );
/*API*/			device.dac().writeVolts( TEST_CHANNEL, volts );
				System.out.println( volts + " volts (" + ( int ) device.dac().voltsToCounts( TEST_CHANNEL, volts ) + " D/A counts)"
					+ " successfully output to channel " + TEST_CHANNEL );

				/*
				 * demonstrate writing volts to multiple D/A channels
				 */
/*API*/			OutputVoltagePoint[] voltPoints = new OutputVoltagePoint[ numChannels ];
				for( int channel = 0; channel < numChannels; channel++ )
/*API*/				voltPoints[ channel ] = new OutputVoltagePoint( channel, ( channel + 1 ) * 10.0 / numChannels );
/*API*/			device.dac().writeVolts( voltPoints );
				System.out.println( "Multiple volts successfully output to " + numChannels + " channels" );
			} else
				System.out.println( "No USB-DA12-8A devices found on USB bus" );
/*API*/		deviceManager.close();
		} catch( Exception ex ) {
			System.err.println( "Error \'" + ex.toString() + "\' occurred while manipulating device" );
			if(
				deviceManager != null
/*API*/			&& deviceManager.isOpen()
			)
/*API*/			deviceManager.close();
		}	// catch( ...
	}	// main()
}	// class Sample

/* end of file */

File: AIOUSB/samples/USB-DA12-8A/sample.cpp
/*
 * $Date $Format: %ad$$
 * $Author $Format: %an <%ae>$$
 * $Release $Format: %h$$
 * AIOUSB library sample program
 */

/*
 * This source code looks best with a tab width of 4.
 *
 * All the API functions that DO NOT begin "AIOUSB_" are standard API functions, largely
 * documented in http://accesio.com/MANUALS/AIOUSB_API_Reference.html. The functions
 * that DO begin with "AIOUSB_" are "extended" API functions added to the Linux
 * implementation. Source code lines in this sample program that are prefixed with the
 * comment "/ * API * /" highlight calls to the AIOUSB API.
 *
 * LIBUSB (http://www.libusb.org/) must be installed on the Linux box (the AIOUSB code
 * was developed using libusb version 1.0.3). After installing libusb, it may also be
 * necessary to set an environment variable so that the libusb and aiousb header files can
 * be located:
 *
 *     export CPATH=/usr/local/include/libusb-1.0/:/usr/local/include/aiousb/
 *
 * Once libusb is installed properly, it should be possible to compile the sample program
 * using the simple command:
 *
 *    make sample AIOUSBLIBDIR=${AIO_LIB_DIR} AIOUSBCLASSLIBDIR=${AIO_CLASSLIB_DIR} DEBUG=1
 *
 */


#include <aiousb.h>
#include <stdio.h>
using namespace AIOUSB;


int 
main( int argc, char **argv ) {
    printf( "USB-DA12-8A/E sample program version 1.112, 12 January 2020\n"
            "  AIOUSB library version %s, %s\n"
            "  This program demonstrates controlling a USB-DA12-8A/E device on\n"
            "  the USB bus. For simplicity, it uses the first such device found\n"
            "  on the bus.\n", AIOUSB_GetVersion(), AIOUSB_GetVersionDate()
            );

    /*
     * MUST call AIOUSB_Init() before any meaningful AIOUSB functions;
     * AIOUSB_GetVersion() above is an exception
     */
    unsigned long result = AIOUSB_Init();
    if( result == AIOUSB_SUCCESS ) {
        /*
         * call GetDevices() to obtain "list" of devices found on the bus
         */
       	unsigned long deviceMask = GetDevices();
        if( deviceMask != 0 ) {
            /*
             * at least one ACCES device detected, but we want one of a specific type
             */
            AIOUSB_ListDevices();				// print list of all devices found on the bus
             int MAX_NAME_SIZE = 20;
            char name[ MAX_NAME_SIZE + 2 ];
            unsigned long productID, nameSize, numDIOBytes, numCounters;
            unsigned long deviceIndex = 0;
            bool deviceFound = false;
            while( deviceMask != 0 ) {
                if( ( deviceMask & 1 ) != 0 ) {
                    nameSize = MAX_NAME_SIZE;
                    result = QueryDeviceInfo( deviceIndex, &productID, &nameSize, name, &numDIOBytes, &numCounters );
                    if( result == AIOUSB_SUCCESS ) {

                        if(
                           productID >= 0x4002
                           && productID <= 0x4003
                           ) {
                            // found a USB-DA12-8A/E family device
                            deviceFound = true;
                            break;				// from while()
                        }
                    } else
                        printf( "Error '%s' querying device at index %lu\n", 
                                AIOUSB_GetResultCodeAsString( result ), deviceIndex );
                }
                deviceIndex++;
                deviceMask >>= 1;
            }
            if( deviceFound ) {
                printf("Using Device Index %d\n", deviceIndex);
                AIOUSB_SetCommTimeout( deviceIndex, 500 );

                uint64_t serialNumber;
                result = GetDeviceSerialNumber( deviceIndex, &serialNumber );
                if( result == AIOUSB_SUCCESS )
                    printf( "Serial number of device at index %lu: %llx\n", 
                            deviceIndex, 
                            ( long long ) serialNumber );
                else
                    printf( "Error '%s' getting serial number of device at index %lu\n", 
                            AIOUSB_GetResultCodeAsString( result ), deviceIndex );

                /*
                 * demonstrate getting enhanced device properties
                 */
                DeviceProperties properties;
                result = AIOUSB_GetDeviceProperties( deviceIndex, &properties );
                if( result == AIOUSB_SUCCESS )
                    printf( "Device properties successfully retrieved\n" );
                else {
                    int MAX_CHANNELS = 8;
                    properties.DACChannels = MAX_CHANNELS;
                    printf( "Error '%s' getting device properties\n"
                            , AIOUSB_GetResultCodeAsString( result ) );
                }

                /*
                 * demonstrate writing to one D/A channel
                 */
                int TEST_CHANNEL = 0;			// channels are numbered 0 to properties.DACChannels - 1
                unsigned MAX_COUNTS = 0xfff;
                unsigned counts = MAX_COUNTS / 2;	// half of full scale
                result = DACDirect( deviceIndex, TEST_CHANNEL, counts );
                if( result == AIOUSB_SUCCESS )
                    printf( "%u D/A counts successfully output to channel %d\n", counts, TEST_CHANNEL );
                else
                    printf( "Error '%s' attempting to output %u D/A counts successfully to channel %d\n"
                            , AIOUSB_GetResultCodeAsString( result )
                            , counts, TEST_CHANNEL );

                /*
                 * demonstrate writing to multiple D/A channels
                 */
                unsigned short dacData[ properties.DACChannels * 2 ];		// channel/count pairs
                for( int channel = 0; channel < ( int ) properties.DACChannels; channel++ ) {
                    dacData[ channel * 2 ] = channel;
                    dacData[ channel * 2 + 1 ] = ( unsigned short ) (
                                                                     ( unsigned long ) ( channel + 1 )
                                                                     * ( unsigned long ) MAX_COUNTS
                                                                     / properties.DACChannels
                                                                     );
                }	// for( int channel ...
                result = DACMultiDirect( deviceIndex, dacData, properties.DACChannels );
                if( result == AIOUSB_SUCCESS )
                    printf( "D/A counts successfully output to %u channels simultaneously\n", properties.DACChannels );
                else
                    printf( "Error '%s' attempting to output D/A counts to %u channels simultaneously\n"
                            , AIOUSB_GetResultCodeAsString( result )
                            , properties.DACChannels );
            } else
                printf( "Failed to find USB-DA12-8A device\n" );
        } else
            printf( "No ACCES devices found on USB bus\n" );

        /*
         * MUST call AIOUSB_Exit() before program exits,
         * but only if AIOUSB_Init() succeeded
         */
       	AIOUSB_Exit();
    }
    return ( int ) result;
}
